tools/fast/fasting/basic.go

```diff
@@ -0,0 +1,132 @@
+
+	err := series.InitAndStart(env.ctx, p)
+	require.NoError(err)
+
+	return p
+}
+
+func (env *Environment) NewNodeConnected() *fast.Filecoin {
+	require := require.New(env.t)
+
+	p := env.NewNodeStarted()
+
+	err := series.Connect(env.ctx, env.GenesisMiner, p)
+	require.NoError(err)
+
+	return p
+}
+
+func (env *Environment) NewNodeWithFunds(funds int) *fast.Filecoin {
+	require := require.New(env.t)
+
+	p := env.NewNodeConnected()
+
+	err := series.SendFilecoinDefaults(env.ctx, env.GenesisMiner, p, funds)
+	require.NoError(err)
+
+	return p
+}
+
+// BasicFastSetup creates a environment and a single node, and environment options
+func BasicFastSetup(ctx context.Context, t *testing.T, fastenvOpts fast.EnvironmentOpts) (*Environment, context.Context) {
+	require := require.New(t)
+
+	// Create a directory for the test using the test name (mostly for FAST)
+	// Replace the forward slash as tempdir can't handle them
+	dir, err := ioutil.TempDir("", strings.Replace(t.Name(), "/", ".", -1))
+	require.NoError(err)
+
+	// Create an environment that includes a genesis block with 1MM FIL
+	env, err := fast.NewEnvironmentMemoryGenesis(big.NewInt(1000000), dir)
+	require.NoError(err)
+
+	// Setup options for nodes.
+	options := make(map[string]string)
+	options[localplugin.AttrLogJSON] = "1"                                        // Enable JSON logs
+	options[localplugin.AttrLogLevel] = "5"                                       // Set log level to Debug
+	options[localplugin.AttrUseSmallSectors] = "true"                             // Enable small sectors
+	options[localplugin.AttrFilecoinBinary] = testhelpers.MustGetFilecoinBinary() // Get the filecoin binary
+
+	genesisURI := env.GenesisCar()
+	genesisMiner, err := env.GenesisMiner()
+	require.NoError(err)
+
+	fastenvOpts = fast.EnvironmentOpts{
+		InitOpts:   append([]fast.ProcessInitOption{fast.POGenesisFile(genesisURI)}, fastenvOpts.InitOpts...),
+		DaemonOpts: append([]fast.ProcessDaemonOption{fast.POBlockTime(time.Millisecond)}, fastenvOpts.DaemonOpts...),
+	}
+
+	// Setup the first node which is used to help coordinate the other nodes by providing
+	// funds, mining for the network, etc
+	genesis, err := env.NewProcess(ctx, localplugin.PluginName, options, fastenvOpts)
+	require.NoError(err)
+
+	err = series.SetupGenesisNode(ctx, genesis, genesisMiner.Address, files.NewReaderFile(genesisMiner.Owner))
+	require.NoError(err)
+
+	// Define a MiningOnce function which will bet set on the context to provide
+	// a way to mine blocks in the series used during testing
+	var MiningOnce series.MiningOnceFunc = func() {
+		_, err := genesis.MiningOnce(ctx)
+		require.NoError(err)
+	}
+
+	ctx = series.SetCtxMiningOnce(ctx, MiningOnce)
+	ctx = series.SetCtxSleepDelay(ctx, time.Second)
+
+	return &Environment{
+		Environment:  env,
+		t:            t,
+		ctx:          ctx,
+		pluginName:   localplugin.PluginName,
+		pluginOpts:   options,
+		fastenvOpts:  fastenvOpts,
+		GenesisMiner: genesis,
```

Comment:
BLOCKING: Maybe this is a different thing, but I think the Environment [already has this value](https://github.com/filecoin-project/go-filecoin/blob/1968f9a3d5c27f122ddcf48e404504b4c5d84050/tools/fast/fasting/basic.go#L31)? As well as the others, with the exception of `testing.T`. Lets talk about this more before merging.

Reply:
The linked environment is the struct that this is filling out. Did you mean on the [`environment` interface?

https://github.com/filecoin-project/go-filecoin/blob/e3f7439e2ac63e555820a8ec7e0b290974042786/tools/fast/environment.go#L36-L46

The value returned by GenesisMiner function is not a filecoin node, but struct representing the miner in the genesis block with power, basically an account actor / wallet private key with the address of the miner in the block.

Reply:
Ahh I see what's happening now, I got confused by the naming, all good here.